### PR TITLE
Make ts-input for form elements

### PR DESCRIPTION
**Pull request by zdlm**
It is only some css on the input. I will going to make storybook and happo story when we have ts-from which includes all the elements we have. 
Screenshot 2020-01-09 at 15 25 34

@esmailpour-hosein @kirill-kolombet @aleks-elkin

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -107,6 +109,21 @@ <h2>Status</h2>
 				</div>
 			</article>
 
+			<article>
+				<h2>Form</h2>
+				<ts-label>Text</ts-label>
+				<ts-input id="test_input">
+					<input />
+				</ts-input>
+				<!-- <script>
```
Comment (tradeshiftci) [inline index.html]: MAJOR Remove this commented out code. [rule](https://sonar.ts.sv/coding_rules#rule_key=Web%3AAvoidCommentedOutCodeCheck)

File: packages/components/input/src/input.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+			if ((this.disabled !== undefined || this.disabled !== null) && input) {
+				input.disabled = true;
+			}
+		}
+
+		getSlotInput() {
+			const slot = this.shadowRoot.querySelector('slot');
+			const elements = slot.assignedNodes().filter(el => el.constructor.name === 'HTMLInputElement');
+			if (elements.length < 1) {
+				console.warn('We need one input element in the slot');
```
Comment (tradeshiftci) [inline packages/components/input/src/input.js]: MINOR Remove this logging statement. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS2228)

File: packages/components/input/src/input.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			}
+			if (elements.length > 1) {
+				console.warn('We need only one input element in the slot');
```
Comment (tradeshiftci) [inline packages/components/input/src/input.js]: MINOR Remove this logging statement. [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AS2228)

File: packages/components/input/src/input.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+		attributeChangedCallback(name, oldval, newval) {
+			const input = this.getSlotInput();
+			if (name === 'disabled' && input) {
+				input.disabled = newval;
+			}
+			super.attributeChangedCallback(name, oldval, newval);
+		}
+
+		firstUpdated(changeedProperties) {
```
Comment (tradeshiftci) [inline packages/components/input/src/input.js]: MAJOR Remove the unused function parameter "changeedProperties". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: packages/components/input/src/input.css [hunk 5]
```diff
@@ -0,0 +1,37 @@
+/* General........................................................ */
+
+:host,
+:host > div {
+	display: block;
```
Comment (kirill-kolombet) [inline packages/components/input/src/input.css]: The div by default is a block-level element, can you check please if this line required?

File: packages/components/input/src/input.css [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	background-color: var(--ts-color-white);
+	white-space: nowrap;
+	word-wrap: normal;
+	overflow: hidden;
+	text-overflow: ellipsis;
+	box-sizing: border-box;
+}
+
+::slotted(input:not([disabled]):hover) {
+	border-color: var(--ts-color-blue-light);
```
Comment (kirill-kolombet) [inline packages/components/input/src/input.css]: In Figma design hover state of _ts-input_ borders are 2 pixels thick. Can you check it out, please?

File: packages/components/input/src/input.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+			return [TSElement.styles, unsafeCSS(css)];
+		}
+
+		static get properties() {
+			return {
+				disabled: { type: Boolean, attribute: 'disabled', reflect: true }
+			};
+		}
+
+		attributeChangedCallback(name, oldval, newval) {
```
[the remark this answers is not in the document]
Comment (zdlm) [reply packages/components/input/src/input.js]: 👍

File: packages/components/label/src/label.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import { TSElement, unsafeCSS, html, customElementDefineHelper } from '@tradeshift/elements';
+import css from './label.css';
+
+customElementDefineHelper(
+	'ts-label',
```
[the remark this answers is not in the document]
Comment (zdlm) [reply packages/components/label/src/label.js]: Yes, That is true.

File: packages/components/input/src/input.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+			if (name === 'disabled' && input && newval !== undefined && newval !== null) {
+				input.disabled = true;
+			}
+			super.attributeChangedCallback(name, oldval, newval);
+		}
+
+		firstUpdated(changeedProperties) {
+			const input = this.getSlotInput();
+			if (this.disabled !== undefined && this.disabled !== null && input) {
+				input.disabled = true;
```
[the remark this answers is not in the document]
Comment (zdlm) [reply packages/components/input/src/input.js]: Good job. Thanks

File: packages/components/label/src/label.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+	class extends TSElement {
+		static get styles() {
+			return [TSElement.styles, unsafeCSS(css)];
+		}
+
+		render() {
+			return html`
+				<label>
+					<slot></slot>
```
[the remark this answers is not in the document]
Comment (zdlm) [reply packages/components/label/src/label.js]: 👍

File: packages/components/input/src/input.css (hunk 5 above)
Comment (zdlm) [reply packages/components/input/src/input.css]: 👍

File: packages/components/input/src/input.css (hunk 6 above)
Comment (zdlm) [reply packages/components/input/src/input.css]: Nice catch

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: packages/components/label/src/label.js [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+			return [TSElement.styles, unsafeCSS(css)];
+		}
+
+		static get properties() {
+			return {
+				for: { type: String, reflect: true }
+			};
+		}
+
+		handleClick(e) {
```
Comment (tradeshiftci) [inline packages/components/label/src/label.js]: MAJOR Remove the unused function parameter "e". [rule](https://sonar.ts.sv/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: index.html (hunk 1 above)
Comment (esmailpour-hosein) [reply index.html]: Can you please remove this?

File: package-lock.json [hunk 12]
```diff
@@ -5782,142 +5820,757 @@
...
+        "@webcomponents/webcomponents-platform": {
+          "version": "1.0.1",
+          "resolved": "https://registry.npmjs.org/@webcomponents/webcomponents-platform/-/webcomponents-platform-1.0.1.tgz",
+          "integrity": "sha512-7Ua2c3FvqAuiC2++gIoStG9r0B5sxleq8B7o0XKuIM052amWPTaCka0UMvnQRRJEsdGgRGIUv6sLkOyfhcr1dw=="
+        },
+        "loadash": {
+          "version": "1.0.0",
+          "resolved": "https://registry.npmjs.org/loadash/-/loadash-1.0.0.tgz",
+          "integrity": "sha512-xlX5HBsXB3KG0FJbJJG/3kYWCfsCyCSus3T+uHVu6QL6YxAdggmm3QeyLgn54N2yi5/UE6xxL5ZWJAAiHzHYEg=="
+        }
```
Comment (esmailpour-hosein) [inline package-lock.json]: Can you please make sure these are not in the package-lock?

File: package-lock.json [hunk 13]
```diff
@@ -5782,142 +5820,757 @@
...
+          "version": "0.7.5",
+          "resolved": "https://registry.npmjs.org/@tradeshift/elements/-/elements-0.7.5.tgz",
+          "integrity": "sha512-0j8L6e4iqgPH9D6MB2vh9fAzdZCZlMCAS/+J8+kWIeaBmWbFC8xpP0y0rzhSZovqvlaBZP5PiLaY4Qr3Mu6gpQ==",
+          "requires": {
+            "lit-element": "^2.2.1"
+          }
+        },
+        "@tradeshift/elements.button": {
+          "version": "0.5.0",
+          "resolved": "https://npm.tradeshift.net/repository/npm-all/@tradeshift/elements.button/-/elements.button-0.5.0.tgz",
```
Comment (esmailpour-hosein) [inline package-lock.json]: This should point to the npm global registry, not tradeshift.

File: packages/components/input/README.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    <a href="https://npmcharts.com/compare/@tradeshift/elements.input?minimal=true">
+		  <img alt="Downloads per month" src="https://badgen.net/npm/dm/@tradeshift/elements.input" height="20"/>
+		</a>
+		<a href="https://www.npmjs.com/browse/depended/@tradeshift/elements.input">
+		  <img alt="Dependent packages" src="https://badgen.net/npm/dependents/@tradeshift/elements.input" height="20"/>
+		</a>
+</p>
+
+## ➤ How to use it
+
```
Comment (esmailpour-hosein) [inline packages/components/input/README.md]: Can you please add slots, props and events here? Also please add how they should be used (adding the input as a slotted node) and the reason for it.

File: package.json [hunk 15]
```diff
@@ -155,7 +157,7 @@
     "rollup-plugin-node-resolve": "^4.2.3",
     "rollup-plugin-postcss": "^2.0.3",
     "rollup-plugin-svgo": "^1.0.3",
-    "rollup-plugin-terser": "^4.0.4"
+    "rollup-plugin-terser": "^5.2.0"
```
Comment (esmailpour-hosein) [inline package.json]: Why did you need to update this?

Comment (zdlm) [reply package.json]: No specific reason. I saw a warning in items to ask for updating.

File: packages/components/input/src/input.js [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+			return {
+				disabled: { type: Boolean, attribute: 'disabled', reflect: true }
+			};
+		}
+
+		attributeChangedCallback(name, oldVal, newVal) {
+			const input = this.getSlotInput();
+			if (name === 'disabled' && input) {
+				input.disabled = newVal !== undefined && newVal !== null;
+			}
```
Comment (esmailpour-hosein) [inline packages/components/input/src/input.js]: What's the reason for this?
Why do we need the disabled state inside the component, if it's only for styling

Comment (zdlm) [reply packages/components/input/src/input.js]: I was thinking to support disabled status both on the ts-input and the input in the slot. 
May be we need to support disabled form that all the elements in the form should be disabled. 
This if or sync the disable status from the host to slot.

File: package-lock.json (hunk 12 above)
Comment (esmailpour-hosein) [reply package-lock.json]: These are still here 🤔

**Comment by tradeshiftci** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'
